testament/testament.nim

```diff
@@ -1220,155 +1497,908 @@ proc parseArgs(execState: var Execution, p: var OptParser): ParseCliResult =
+            test: baseTest,
+            target: target,
+            matrixEntry: noMatrixEntry)
+
+          run.test.name = testFiles[testId].makeName(
+            run, spec.err == reKnownIssue)
+
+          execState.testRuns.add run
+          execState.runTimes.add RunTime()
+          execState.runActuals.add RunActual()
+
+        else:
+          for entryId, _ in spec.matrix.pairs:
+            var run = TestRun(
+              testId: testId,
+              target: target,
+              matrixEntry: entryId,
+              test: baseTest
+            )
+
+            run.test.options = spec.matrix[entryId]
+            run.test.name = testFiles[testId].makeName(
+              run, spec.err == reKnownIssue)
+
+            execState.testRuns.add run
+            execState.runTimes.add RunTime()
+            execState.runActuals.add RunActual()
+
+proc prepareTestActions(execState: var Execution) =
+  ## create a list of necessary test actions
+  # IMPLEMENT : create specific type to avoid accidental mutation
+
+  # NOTE: these are what are read, so a dedicate type would offer a
+  # read-only view of those, but allow actions mutation
+  let
+    testRuns = execState.testRuns
+    testSpecs = execState.testSpecs
+
+  # TODO: handle disabled and known issue
+
+  for runId, run in testRuns.pairs:
+    let actionKind = testSpecs[run.testId].action
+    case actionKind
+    of actionReject, actionCompile:
+      execState.actions.add:
+        TestAction(runId: runId, kind: actionKind)
+    of actionRun:
+      let compileActionId = execState.actions.len
+      execState.actions.add:
+        TestAction(runId: runId, kind: actionCompile, partOfRun: true)
+      execState.actions.add:
+        TestAction(runId: runId,
+                   kind: actionRun,
+                   compileActionId: compileActionId)
+
+proc reportTestRunResult(
+    legacyResults: var TResults,
+    execution: Execution,
+    cat: Category,
+    testFile: TestFile,
+    spec: TSpec,
+    testRun: TestRun,
+    runActual: RunActual,
+    runTime: RunTime,
+    action: TestAction,
+    cmd: string,
+    debugInfo: string
+  ) =
+  ## Report results of a single test run
+  # REFACTOR birdge newer testament internals (`ExecutionState` and
+  # friends) to the legacy reporting and generate output accordingly.
+
+  let
+    duration = runTime.compileEnd - runTime.compileStart +
+                runTime.runEnd - runTime.runStart
+    target = testRun.target
+    allowFailure = spec.err == reKnownIssue
 
-  ## Main procedure
-  backend.open()
-  ## Cli options and misc
   var
-    optPrintResults = false
-    isMainProcess   = true
-    skipFrom        = ""
-  
-  ## First parse options
-  case parseOpts(execState, p)
-  of parseQuitWithUsage: quit Usage
-  of parseSuccess: discard
+    # REFACTOR test run should contianer ID of the specification it was
+    # used for as well as information about it's matrix parameters. The ID
+    # should be used instead of this copy.
```

Comment:
```suggestion
    # REFACTOR test run should contian the ID of the specification it's
    # associated with as well as information about its matrix parameters.
    # The ID should be used instead of this copy.
```

Reply:
Although, that might not be enough and it might have to be at the action level.